src/dialogue_system.gd

```diff
@@ -0,0 +1,379 @@
+  if new_dialogue_name != null and new_dialogue_name != current_dialogue_name:
+    dialogue = load_dialogue(new_dialogue_name) # sets current_dialogue_name and dialogue
+  # Continue to the next block
+  current_block = dialogue[next_block_name]
+  current_block.id = current_dialogue_name + ":" + next_block_name
+  if current_block.has("set_var"):
+    set_variables(current_block["set_var"])  
+  while current_block.has("condition"): # drill down to a non-conditional block
+    var outcome_block = condition_outcome(current_block)
+    if condition_holds(current_block["condition"]):
+      outcome_block.id = current_block.id + ":true"
+    else:
+      outcome_block.id = current_block.id + ":false"
+    current_block = outcome_block
+    if current_block.has("set_var"):
+      set_variables(current_block["set_var"]) 
+  is_question = false
+  if current_block.has("options"):
+    is_question = true
+
+func condition_outcome(obj):
+  if condition_holds(obj["condition"]):
+    return obj.get("true")
+  else:
+    return obj.get("false")
+
+func condition_holds(condition, value_to_match = true):
+  if typeof(condition) == TYPE_STRING:
+    # var_name or dict_name<separator>var_name
+    var dict_and_var = condition.split(config.condition_separator)
+    if len(dict_and_var) == 1: # use default "variables" dictionary
+      return progress.get("variables").get(condition, false) == value_to_match
+    elif len(dict_and_var) == 2: # dictionary is specified
+      return progress.get(dict_and_var[0]).get(dict_and_var[1], false) == value_to_match
+  elif typeof(condition) == TYPE_DICTIONARY:
+    return condition_holds(condition["var_id"], condition["value"])
+  elif typeof(condition) == TYPE_ARRAY: # multiple conditions
+    for subcondition in condition:
+      if not condition_holds(subcondition):
+        return false
+    return true
+
+func filter_options(proto_options):
+  var real_options = []
+  for protopt in proto_options:
+    if typeof(protopt) == TYPE_ARRAY:
+      real_options.append(protopt)
+    elif typeof(protopt) == TYPE_DICTIONARY:
+      var conditional = condition_holds(protopt["condition"])
+      if conditional and protopt.has("true"):
+        real_options.append(protopt["true"])
+      elif not conditional and protopt.has("false"):
+        real_options.append(protopt["false"])
+    else:
+      print("ERROR: option has unexpected type " + str(typeof(protopt)))
+  return real_options
+
+func load_dialogue(file_id):
+  # Load new dialogue
+  current_dialogue_name = file_id
+  var file = File.new()
+  file.open('%s/%s.json' % [config.dialogues_folder, current_dialogue_name], file.READ)
+  var json = file.get_as_text()
+  file.close()
+  var parsed_json = JSON.parse(json)
+  if len(parsed_json.get("error_string")) > 0:
+    print(parsed_json["error_string"])
+    print("Error on line " + str(parsed_json["error_line"]))
+  dialogue = parsed_json.result
+  return dialogue
+
+func set_variables(assignment_arrays):
```

Comment:
It's a small change, but function would read a little better if the unboxing of values was separated from processing and emitting.

So something like this:
```
func set_variables(assignment_arrays):
  for assignment in assignment_arrays:
    var var_group, var_name, var_value

    if len(assignment) == 2: # default variables array
      var_dict = "variables"
      var_name = assignment[0]
      var_value = assignment[1]
    else: # assignment = [dictionary, key, value]
      var_dict = assignment[0]
      var_name = assignment[1] 
      var_value = assignment[2]

    progress.get(var_dict)[var_name] = value
    emit_signal("GIFT_var_change", [var_dict, var_name, value])
```